processing/src/main/java/org/apache/druid/math/expr/Expr.java

```diff
@@ -763,4 +763,36 @@ private static Set<String> map(
       return results;
     }
   }
+
+
+  /**
+   * Returns the single-threaded version of the given expression tree.
+   *
+   * Nested expressions in the subtree are also optimized.
+   * Individual {@link Expr}-s which have a singleThreaded implementation via {@link SingleThreaded} are substituted.
+   */
+  static Expr singleThreaded(Expr expr)
+  {
+    return expr.visit(
+        node -> {
+          if (node instanceof SingleThreaded) {
+            SingleThreaded canBeSingleThreaded = (SingleThreaded) node;
+            return canBeSingleThreaded.toSingleThreaded();
+          } else {
+            return node;
+          }
+        }
+    );
+  }
+
+  /**
+   * Implementing this interface allows to provide a non-threadsafe {@link Expr} implementation.
+   */
+  interface SingleThreaded
```

Comment:
The name here is misleading, since it's not that the Expr _is_ single-threaded, it's that it has a faster single-threaded implementation available. Maybe a better name is: `SingleThreadSpecializable`